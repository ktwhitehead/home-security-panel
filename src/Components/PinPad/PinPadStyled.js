import styled from 'styled-components';

const Body = styled.div`
    display: flex;
    flex-direction: column;
    width: 100%;
`;

const Close = styled.div`
    color: ${props => props.theme.colors.light.primary};
    font-size: 28px;
    font-weight: bold;
    padding: 50px;
    font-size: 4em;
    line-height: 0.4;

    &:hover {
        cursor: pointer;
    }
`;

const Rows = styled.div`
    display: flex;
    justify-content: center;
    flex-direction: column;
    align-items: center;
`;

const Pin = styled.div`
    width: 100%;
    min-height: 5.5em;
    display: flex;
    align-items: center;
    justify-content: center;
`;

const Number = styled.div`
    padding: ${props => props.type === 'pad' ? '0.4em 1em' : '0 0.7em'};
    color: ${props => props.theme.colors.light.primary};
    font-size: 4em;
    font-weight: ${props => props.type === 'pad' ? 'bold' : 'initial'}
`;

const Message = styled.div`
    color: ${props => props.theme.colors.alert};
    align-self: flex-end;
    font-size: 2em;
`;

const Wrapper = styled.div`
    display: ${props => props.visible ? '' : 'none'};
    position: fixed;
    z-index: 1000;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    overflow: auto;
    background-color: ${props => props.theme.colors.dark.primary};
`;

const Row = styled.div`
    display: flex;
`

const Header = styled.div`
    display: flex;
    justify-content: space-between;
`;

export { Body, Close, Pin, Row, Rows, Number, Wrapper, Header, Message };
