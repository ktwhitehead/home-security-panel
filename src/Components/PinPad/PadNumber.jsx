import React, { useContext } from 'react';
import { Number } from './PinPadStyled';
import AppContext from '../../Context/AppContext';

const PadNumber = ({number}) => {
    const { pin, setPin } = useContext(AppContext);
    let newPin = pin.toString() + number.toString();
    return <Number type={'pad'} onClick={() => setPin(newPin)}>{number}</Number>
}

export default PadNumber;