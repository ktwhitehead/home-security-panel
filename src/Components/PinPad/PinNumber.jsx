import React from 'react';
import { Number } from './PinPadStyled';

const PinNumber = ({number}) => {
    return <Number data-testid={`PinNumber-${number}`}>{number}</Number>
}

export default PinNumber;