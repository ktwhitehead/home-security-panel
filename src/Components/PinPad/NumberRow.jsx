import React from 'react';
import PadNumber from './PadNumber';
import { Row } from './PinPadStyled';

const NumberRow = ({numbers}) => {
    return (
        <Row>
            {numbers.map(number => <PadNumber number={number} />)}
        </Row>
    )
}

export default NumberRow;
