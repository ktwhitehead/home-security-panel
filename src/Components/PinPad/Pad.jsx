import React from 'react';
import NumberRow from './NumberRow';
import { Rows } from './PinPadStyled';

const Pad = ({ pin, setPin }) => {
  return (
    <Rows>
      <NumberRow numbers={[1,2,3]} pin={pin} setPin={setPin} />
      <NumberRow numbers={[4,5,6]} pin={pin} setPin={setPin} />
      <NumberRow numbers={[7,8,9]} pin={pin} setPin={setPin} />
      <NumberRow numbers={[0]} pin={pin} setPin={setPin} />
    </Rows>
  )
}

export default Pad;
