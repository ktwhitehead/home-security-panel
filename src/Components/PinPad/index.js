import PinPad from './PinPad';

export default PinPad;
