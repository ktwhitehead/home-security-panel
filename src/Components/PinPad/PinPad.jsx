import React, { useEffect, useState } from 'react';
import { Wrapper, Close, Body, Pin, Header, Message } from './PinPadStyled';
import PinNumber from './PinNumber';
import Pad from './Pad';

const PinPad = ({ display, setDisplay, errorMessage, setErrorMessage, onPinEntered }) => {
  const [pin, setPin] = useState('');

  useEffect(() => {
    if (pin.length === 1) setErrorMessage();
    if (pin.length >= 4) {
      onPinEntered(pin);
      setPin('');
    }
  }, [pin]);

  const closePinPad = () => {
    setDisplay(false);
    setErrorMessage();
  };

  return (
    <Wrapper visible={display} data-testid="PinPad">
      <Header>
        <Close style={{visibility: 'hidden'}}>x</Close>
        <Message data-testid="PinPadError">{errorMessage}</Message>
        <Close onClick={closePinPad} data-testid="ClosePinPad">x</Close>
      </Header>
      <Body>
        <Pin data-testid={"EnteredPinNumber"}>
          {pin.split('').map((number, i) => <PinNumber key={`pin-number-${number}-${i}`} number={number} />)}
        </Pin>
        <Pad pin={pin} setPin={setPin} />
      </Body>
    </Wrapper>
  );
};

export default PinPad;