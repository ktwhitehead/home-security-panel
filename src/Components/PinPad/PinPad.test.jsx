import React from 'react';
import PinPad from './PinPad';
import withContext from '../../Test';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'

describe('PinPad', () => {
  test('displays a pin pad', () => {
    const setDisplay = jest.fn();
    const onPinEntered = jest.fn();
    const { getByTestId } = render(withContext(<PinPad display setDisplay={setDisplay} onPinEntered={onPinEntered} />));

    const pinPad = getByTestId('PinPad');
    expect(pinPad).toBeVisible();
  });

  test('displays the pin number entered', () => {
    const setDisplay = jest.fn();
    const onPinEntered = jest.fn();
    const setErrorMessage = jest.fn();
    const { getByTestId } = render(
      withContext(<PinPad display setDisplay={setDisplay} onPinEntered={onPinEntered} setErrorMessage={setErrorMessage} />)
    );

    const padNumber1 = getByTestId('PadNumber-1');
    const padNumber9 = getByTestId('PadNumber-9');

    fireEvent.click(padNumber1);
    fireEvent.click(padNumber9);

    const enteredNumber1 = getByTestId('PinNumber-1');
    const enteredNumber9 = getByTestId('PinNumber-9');

    expect(enteredNumber1).toBeVisible();
    expect(enteredNumber9).toBeVisible();
  });

  test('removes the pin numbers when 4 numbers have been entered', () => {
    const setDisplay = jest.fn();
    const onPinEntered = jest.fn();
    const setErrorMessage = jest.fn();
    const { getByTestId, queryByTestId } = render(
      withContext(<PinPad display setDisplay={setDisplay} onPinEntered={onPinEntered} setErrorMessage={setErrorMessage} />)
    );

    const padNumber1 = getByTestId('PadNumber-1');
    const padNumber2 = getByTestId('PadNumber-2');
    const padNumber3 = getByTestId('PadNumber-3');
    const padNumber4 = getByTestId('PadNumber-4');

    fireEvent.click(padNumber1);
    fireEvent.click(padNumber2);
    fireEvent.click(padNumber3);
    fireEvent.click(padNumber4);

    expect(onPinEntered).toHaveBeenCalledWith('1234');

    const enteredNumber1 = queryByTestId('PinNumber-1');
    const enteredNumber2 = queryByTestId('PinNumber-2');
    const enteredNumber3 = queryByTestId('PinNumber-3');
    const enteredNumber4 = queryByTestId('PinNumber-4');

    expect(enteredNumber1).toBeNull();
    expect(enteredNumber2).toBeNull();
    expect(enteredNumber3).toBeNull();
    expect(enteredNumber4).toBeNull();
  });

  test('displays an error message if provided', () => {
    const setDisplay = jest.fn();
    const onPinEntered = jest.fn();
    const errorMessage = 'Error Message'
    const { getByTestId } = render(
      withContext(<PinPad display setDisplay={setDisplay} onPinEntered={onPinEntered} errorMessage={errorMessage} />)
    );

    expect(getByTestId('PinPadError').innerHTML).toEqual(errorMessage);
  });

  test('matches the expected styling', () => {
    const setDisplay = jest.fn();
    const onPinEntered = jest.fn();
    const { getByTestId } = render(withContext(<PinPad display setDisplay={setDisplay} onPinEntered={onPinEntered} />));

    const pinPad = getByTestId('PinPad');
    expect(pinPad).toMatchSnapshot();
  });
});
