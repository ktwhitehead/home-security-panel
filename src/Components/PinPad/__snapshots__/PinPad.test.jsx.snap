// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinPad matches the expected styling 1`] = `
<div
  class="PinPadStyled__Wrapper-sc-1k5gjku-6 knRLMV"
  data-testid="PinPad"
>
  <div
    class="PinPadStyled__Header-sc-1k5gjku-8 eaZouw"
  >
    <div
      class="PinPadStyled__Close-sc-1k5gjku-1 dtzUzz"
      style="visibility: hidden;"
    >
      x
    </div>
    <div
      class="PinPadStyled__Message-sc-1k5gjku-5 jEQlPV"
      data-testid="PinPadError"
    />
    <div
      class="PinPadStyled__Close-sc-1k5gjku-1 dtzUzz"
      data-testid="ClosePinPad"
    >
      x
    </div>
  </div>
  <div
    class="PinPadStyled__Body-sc-1k5gjku-0 kqmgaY"
  >
    <div
      class="PinPadStyled__Pin-sc-1k5gjku-3 dCNkGn"
      data-testid="EnteredPinNumber"
    />
    <div
      class="PinPadStyled__Rows-sc-1k5gjku-2 hgfPBA"
    >
      <div
        class="PinPadStyled__Row-sc-1k5gjku-7 jSQshO"
      >
        <div
          class="PinPadStyled__Number-sc-1k5gjku-4 hQrIYu"
          data-testid="PadNumber-1"
          type="pad"
        >
          1
        </div>
        <div
          class="PinPadStyled__Number-sc-1k5gjku-4 hQrIYu"
          data-testid="PadNumber-2"
          type="pad"
        >
          2
        </div>
        <div
          class="PinPadStyled__Number-sc-1k5gjku-4 hQrIYu"
          data-testid="PadNumber-3"
          type="pad"
        >
          3
        </div>
      </div>
      <div
        class="PinPadStyled__Row-sc-1k5gjku-7 jSQshO"
      >
        <div
          class="PinPadStyled__Number-sc-1k5gjku-4 hQrIYu"
          data-testid="PadNumber-4"
          type="pad"
        >
          4
        </div>
        <div
          class="PinPadStyled__Number-sc-1k5gjku-4 hQrIYu"
          data-testid="PadNumber-5"
          type="pad"
        >
          5
        </div>
        <div
          class="PinPadStyled__Number-sc-1k5gjku-4 hQrIYu"
          data-testid="PadNumber-6"
          type="pad"
        >
          6
        </div>
      </div>
      <div
        class="PinPadStyled__Row-sc-1k5gjku-7 jSQshO"
      >
        <div
          class="PinPadStyled__Number-sc-1k5gjku-4 hQrIYu"
          data-testid="PadNumber-7"
          type="pad"
        >
          7
        </div>
        <div
          class="PinPadStyled__Number-sc-1k5gjku-4 hQrIYu"
          data-testid="PadNumber-8"
          type="pad"
        >
          8
        </div>
        <div
          class="PinPadStyled__Number-sc-1k5gjku-4 hQrIYu"
          data-testid="PadNumber-9"
          type="pad"
        >
          9
        </div>
      </div>
      <div
        class="PinPadStyled__Row-sc-1k5gjku-7 jSQshO"
      >
        <div
          class="PinPadStyled__Number-sc-1k5gjku-4 hQrIYu"
          data-testid="PadNumber-0"
          type="pad"
        >
          0
        </div>
      </div>
    </div>
  </div>
</div>
`;
